addons/mail/wizard/mail_template_preview.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+LIST_FIELDS_MAIL_TEMPLATE = ['subject', 'body_html', 'email_from', 'email_to', 'partner_to', 'email_cc', 'reply_to', 'scheduled_date', 'attachment_ids']
+
+
+class MailTemplatePreview(models.TransientModel):
+    _inherit = "mail.template"
+    _name = "mail.template.preview"
+    _description = "Email Template Preview"
+
+    @api.model
+    def _selection_target_model(self):
+        return [(model.model, model.name) for model in self.env['ir.model'].search([])]
+
+    @api.model
+    def _selection_languages(self):
+        return self.env['res.lang'].get_installed()
+
+    @api.model
+    def default_get(self, fields):
+        result = super(MailTemplatePreview, self).default_get(fields)
+        default_mail_template_id = self.env.context.get('default_mail_template_id')
+        if not default_mail_template_id or 'resource_ref' not in fields:
+            return result
+        mail_template = self.env['mail.template'].browse(default_mail_template_id)
+        res = self.env[mail_template.model_id.model].search([], limit=1)
+        if res:
+            result['resource_ref'] = '%s,%s' % (mail_template.model_id.model, res.id)
+        return result
+
+    mail_template_id = fields.Many2one('mail.template', string="Related Mail Template")
+    model_id = fields.Many2one('ir.model', string='Targeted model', related="mail_template_id.model_id")
+    resource_ref = fields.Reference(string='Record reference', selection='_selection_target_model')
+    lang = fields.Selection(_selection_languages, string='Template Preview Language')
+    partner_ids = fields.Many2many('res.partner', string='Recipients')
+    # Override fields from mail.template model
+    attachment_ids = fields.Many2many(string='Attachments', store=False)
+    # Error Helper fields
+    no_record = fields.Boolean('No record of this model', compute='_compute_no_record')
+    error_msg = fields.Char('Error Message', readonly=True)
+
+    @api.depends('model_id')
+    def _compute_no_record(self):
+        if self.model_id:
+            self.no_record = self.env[self.model_id.model].search_count([]) == 0
+
+    @api.onchange('lang', 'resource_ref')
+    def on_change_resource_ref(self):
+        """ Preview the mail template (body, subject, ...) depending of the language and
+        the record reference, more precisely the record id for the defined model of the mail template.
+        If no record id is selectable/set, the jinja placeholders won't be replace in the display information.
+        """
+        if not self.resource_ref:
+            for field in LIST_FIELDS_MAIL_TEMPLATE:
```

Comment:
This feels like a lot of repetition.
I've been playing around with the code and came up with this:

```
def on_change_resource_ref(self):
    mail_template = self.mail_template_id.with_context(lang=self.lang)
    try:
        if not self.resource_ref:
            raise UserError(_("No records for this model, fallback to default values"))
        mail_values = mail_template.with_context(template_preview_lang=self.lang).generate_email(self.resource_ref.id, LIST_FIELDS_MAIL_TEMPLATE)
        self._set_mail_attributes(mail_values, True)
        self.error_msg = False
    except UserError user_error:
        self._set_mail_attributes(mail_template)
        self.error_msg = user_error.message

def _set_mail_attributes(self, values, include_partner_ids=False):
    fields = LIST_FIELDS_MAIL_TEMPLATE
    for field in (fields if not include_partner_ids else fields + ['partner_id]):
        if isinstance(values, dict):
            setattr(self, field, values.get(field, False))
        else:
            setattr(self, field, getattr(values, field))
```

This leaves the error message to the generate_email method, which could be more precise.
It would also allow you to get rid of the "no_record" thing.

Let me know what you think about it, maybe I've missed something.